cp_diffdock_protocol.py

```diff
@@ -0,0 +1,121 @@
+            )
+            # Need to preserve defaults for these if not in request
+            if data.get('samples_per_complex') is not None: req.samples_per_complex = data['samples_per_complex']
+            if data.get('add_hs') is not None: req.add_hs = data['add_hs']
+            if data.get('keep_hs') is not None: req.keep_hs = data['keep_hs']
+            if data.get('keep_src_3d') is not None: req.keep_src_3d = data['keep_src_3d']
+            return req
+            
+    @dataclass
+    class Pose:
+        """
+        One pose from a DiffDock result.
+        """
+        filename: str
+        sdf: str
+        rank: int
+        confidence: float
+
+        def __iter__(self):
+            return iter(astuple(self))
+
+    @dataclass
+    class Result:
+        """
+        Results for one DiffDock protein-ligand complex, either a list of poses or an error.
+        """
+        proteinName: str
+        ligandName: str
+        poses: List['DiffDockProtocol.Pose'] = field(default_factory=list)
+        error: str = ''
+
+        @staticmethod
+        def from_dict(data: dict):
+            return DiffDockProtocol.Result(
+                proteinName=data['proteinName'],
+                ligandName=data['ligandName'],
+                poses=[DiffDockProtocol.Pose(**p) for p in data['poses']],
+                error=data['error']
+            )
+
+    @dataclass
+    class Response:
+        """
+        Response for the entire DiffDock request, either a list of results, an error, or a status update.
+        """
+        messageType: 'DiffDockProtocol.MessageType' = 'invalid'
+        results: List['DiffDockProtocol.Result'] = field(default_factory=list)
+        message: str = ''
+
+        @staticmethod
+        def makeError(message: str=""):
+            return DiffDockProtocol.Response(messageType=DiffDockProtocol.MessageType.ERROR, message=message)
+
+        @staticmethod
+        def makeStatus(message: str=""):
+            return DiffDockProtocol.Response(messageType=DiffDockProtocol.MessageType.STATUS, message=message)
+
+        @staticmethod
+        def makeResults(results: List['DiffDockProtocol.Result']=[]):
+            return DiffDockProtocol.Response(messageType=DiffDockProtocol.MessageType.RESULTS, results=results)
+
+        @staticmethod
+        def from_json(data: str):
+            return DiffDockProtocol.Response.from_dict(json.loads(data))
+
+        @staticmethod
+        def from_dict(data: dict):
+            messageType = data['messageType']
+            if messageType not in DiffDockProtocol.MessageType:
+                messageType = DiffDockProtocol.MessageType.INVALID
+            return DiffDockProtocol.Response(
+                messageType=messageType,
+                results=[DiffDockProtocol.Result.from_dict(r) for r in data['results']],
+                message=data['message']
+            )
+
+        def to_json(self):
+            return json.dumps(self, default=lambda x: x.__dict__)
```

Comment:
Would dataclasses.asdict do what getting __dict__ does? If so, it would at least look like it wasn't manipulating internals.

Reply:
Yeah, that is much better